Consider rerunning failed tests

Consider rerunning failed tests to avoid https://travis-ci.org/mockito/mockito/jobs/172673452

Some options:
 - rerun test Gradle task
 - rerun test rule http://stackoverflow.com/questions/8295100/how-to-re-run-failed-junit-tests-immediately

Agreed there's still not so rare build failures, I had to restart the job 177057772 manually. I've been seeng this just as often since #704 merge.

```
Worker information
hostname: travis-worker-gce-org-prod2-4:4d952407-384a-4d90-a194-6a04d5b5e251
version: v2.5.0 https://github.com/travis-ci/worker/tree/da3a43228dffc0fcca5a46569ca786b22991979f
instance: testing-gce-e550a994-d659-4a46-bccf-eed57dafea15:travis-ci-jvm-precise-1471814223
startup: 21.62330554s
Build system information
Build language: java
Build group: stable
Build dist: precise
Build id: 177057771
Job id: 177057772
travis-build version: 0d606872a
Build image provisioning date and time
Sun Aug 21 21:33:22 UTC 2016
Operating System Details
Distributor ID:	Ubuntu
Description:	Ubuntu 12.04.5 LTS
Release:	12.04
Codename:	precise
Linux Version
3.13.0-92-generic
Cookbooks Version
f77e708 https://github.com/travis-ci/travis-cookbooks/tree/f77e708
Git version
git version 1.8.5.6
bash version
GNU bash, version 4.2.25(1)-release (x86_64-pc-linux-gnu)
Copyright (C) 2011 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
GCC version
gcc (Ubuntu/Linaro 4.6.3-1ubuntu5) 4.6.3
Copyright (C) 2011 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

LLVM version
clang version 3.4 (tags/RELEASE_34/final)
Target: x86_64-unknown-linux-gnu
Thread model: posix
Pre-installed Ruby versions
ruby-2.2.5
Pre-installed Node.js versions
v0.10.36
Pre-installed Go versions
1.4.2
mysql --version
mysql  Ver 14.14 Distrib 5.5.50, for debian-linux-gnu (x86_64) using readline 6.2
Pre-installed PostgreSQL versions
9.1.23
9.2.18
9.3.14
9.4.9
Redis version
redis-server 3.0.6
riak version
2.0.2
memcached version
1.4.13
MongoDB version
MongoDB 2.4.14
CouchDB version
couchdb 1.6.1
Neo4j version
1.9.4
Cassandra version
2.0.9
ElasticSearch version
1.4.0
Installed Sphinx versions
2.0.10
2.1.9
2.2.6
Default Sphinx version
2.2.6
Installed Firefox version
firefox 38.4.0esr
PhantomJS version
1.9.8
ant -version
Apache Ant(TM) version 1.8.2 compiled on December 3 2011
mvn -version
Apache Maven 3.2.5 (12a6b3acb947671f09b81f49094c53f426d8cea1; 2014-12-14T17:29:23+00:00)
Maven home: /usr/local/maven
Java version: 1.7.0_80, vendor: Oracle Corporation
Java home: /usr/lib/jvm/java-7-oracle/jre
Default locale: en, platform encoding: UTF-8
OS name: "linux", version: "3.13.0-92-generic", arch: "amd64", family: "unix"
gradle -version

------------------------------------------------------------
Gradle 2.2.1
------------------------------------------------------------

Build time:   2014-11-24 09:45:35 UTC
Build number: none
Revision:     6fcb59c06f43a4e6b1bcb401f7686a8601a1fb4a

Groovy:       2.3.6
Ant:          Apache Ant(TM) version 1.9.3 compiled on December 23 2013
JVM:          1.7.0_80 (Oracle Corporation 24.80-b11)
OS:           Linux 3.13.0-92-generic amd64

lein1 version
WARNING: You're currently running as root; probably by accident.
Press control-C to abort or Enter to continue as root.
Set LEIN_ROOT to disable this warning.
WARNING: You're currently running as root; probably by accident.
Press control-C to abort or Enter to continue as root.
Set LEIN_ROOT to disable this warning.
Downloading Leiningen now...
Using JLine for console I/O; install rlwrap for optimum experience.
Leiningen 1.7.1 on Java 1.7.0_80 Java HotSpot(TM) 64-Bit Server VM
lein2 version
WARNING: You're currently running as root; probably by accident.
Press control-C to abort or Enter to continue as root.
Set LEIN_ROOT to disable this warning.
Downloading Leiningen to /home/travis/.lein/self-installs/home/travis/.lein/leiningen-2.5.1-standalone.jar now...
Leiningen 2.5.1 on Java 1.7.0_80 Java HotSpot(TM) 64-Bit Server VM
lein version
WARNING: You're currently running as root; probably by accident.
Press control-C to abort or Enter to continue as root.
Set LEIN_ROOT to disable this warning.
Downloading Leiningen to /home/travis/.lein/self-installs/home/travis/.lein/leiningen-2.5.1-standalone.jar now...
Leiningen 2.5.1 on Java 1.7.0_80 Java HotSpot(TM) 64-Bit Server VM

$ export DEBIAN_FRONTEND=noninteractive
Reading package lists...
Building dependency tree...
Reading state information...
libc6 is already the newest version.
0 upgraded, 0 newly installed, 0 to remove and 102 not upgraded.
$ git clone --depth=50 --branch=remove-internal-javadoc-links https://github.com/mockito/mockito.git mockito/mockito
Cloning into 'mockito/mockito'...
remote: Counting objects: 3100, done.
remote: Compressing objects: 100% (1762/1762), done.
remote: Total 3100 (delta 1276), reused 2086 (delta 919), pack-reused 0
Receiving objects: 100% (3100/3100), 1.17 MiB | 0 bytes/s, done.
Resolving deltas: 100% (1276/1276), done.
Checking connectivity... done.

$ cd mockito/mockito
$ git checkout -qf 7a4bee38654b052361153cadc1dda5ede068d0ff
Set hostname to mockito-cd
$ sudo hostname mockito-cd

## Managed by Chef on packer-57ba1a4f-8735-ef55-49ff-0a97bf85bb9d.c.eco-emissary-99515.internal :heart_eyes_cat:
## cookbook:: travis_build_environment
##     file:: templates/default/etc/cloud/templates/hosts.tmpl.erb

127.0.1.1 testing-gce-e550a994-d659-4a46-bccf-eed57dafea15 testing-gce-e550a994-d659-4a46-bccf-eed57dafea15 ip4-loopback precise64

sudo: unable to resolve host mockito-cd

## Managed by Chef on packer-57ba1a4f-8735-ef55-49ff-0a97bf85bb9d.c.eco-emissary-99515.internal :heart_eyes_cat:
## cookbook:: travis_build_environment
##     file:: templates/default/etc/cloud/templates/hosts.tmpl.erb

127.0.1.1 testing-gce-e550a994-d659-4a46-bccf-eed57dafea15 testing-gce-e550a994-d659-4a46-bccf-eed57dafea15 ip4-loopback precise64


Setting environment variables from repository settings
$ export NEXUS_TOKEN_USER=[secure]
$ export NEXUS_TOKEN_PWD=[secure]
$ export GH_TOKEN=[secure]
$ export MOCKITO_BINTRAY_API_KEY=[secure]
$ export TEST_ENV_VAR=[secure]

Setting environment variables from .travis.yml
$ export TERM=dumb
$ export GRADLE_OPTS="-XX:+UseCompressedOops -Djava.awt.headless=true"

$ jdk_switcher use openjdk6
Switching to OpenJDK6 (java-1.6.0-openjdk-amd64), JAVA_HOME will be set to /usr/lib/jvm/java-6-openjdk-amd64
$ export TERM=dumb
Setting up build cache
$ export CASHER_DIR=$HOME/.casher
$ Installing caching utilities


attempting to download cache archive
fetching remove-internal-javadoc-links/cache-linux-precise-e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855--jdk-openjdk6.tgz
found cache


adding /home/travis/.gradle/caches to cache
adding /home/travis/.gradle/wrapper to cache

$ java -Xmx32m -version
java version "1.6.0_39"
OpenJDK Runtime Environment (IcedTea6 1.13.11) (6b39-1.13.11-0ubuntu0.12.04.1)
OpenJDK 64-Bit Server VM (build 23.25-b01, mixed mode)
$ javac -J-Xmx32m -version
javac 1.6.0_39
$ export TRAVIS_COMMIT_MESSAGE=$(git log --format=%B -n 1 $TRAVIS_COMMIT)

$ echo "$TRAVIS_COMMIT_MESSAGE"
Disables the testng javadoc task

This task is not configured and shows errors and warning.
Also the test-ng project is not released yet.

$ true

$ ./gradlew ciBuild release -Dscan
Starting a new Gradle Daemon for this build (subsequent builds will be faster).
Support for running Gradle using Java 6 has been deprecated and will be removed in Gradle 3.0
Parallel execution is an incubating feature.
:buildSrc:clean UP-TO-DATE
:buildSrc:compileJava UP-TO-DATE
:buildSrc:compileGroovy UP-TO-DATE
:buildSrc:processResources UP-TO-DATE
:buildSrc:classes UP-TO-DATE
:buildSrc:jar
:buildSrc:assemble
:buildSrc:compileTestJava UP-TO-DATE
:buildSrc:compileTestGroovy UP-TO-DATE
:buildSrc:processTestResources UP-TO-DATE
:buildSrc:testClasses UP-TO-DATE
:buildSrc:test UP-TO-DATE
:buildSrc:check UP-TO-DATE
:buildSrc:build
Version: 2.2.20
:compileJava
:extTest:compileJava UP-TO-DATE
:extTest:processResources UP-TO-DATE
:extTest:classes UP-TO-DATE
:extTest:jar
:extTest:assemble
:extTest:processTestResources
:testng:processResources UP-TO-DATE
:testng:processTestResources
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
:processResources UP-TO-DATE
:classes
:jar
:javadoc
:testng:compileJava
:testng:classes
:testng:jar
:testng:sourcesJar
:testng:assemble
:testng:compileTestJava
:testng:testClasses
:testng:test
:testng:check
:testng:build
Installing sw-precache globally
:javadocJar
:sourcesJar
:assemble
:licenseMain
Unknown file extension: src/main/java/org/mockito/internal/util/concurrent/LICENSE
Unknown file extension: src/main/java/org/mockito/internal/util/concurrent/README.md
:licenseMain UP-TO-DATE
:licenseTest
Missing header in: src/test/java/org/mockitousage/CompilationWarningsTest.java
Missing header in: src/test/java/org/mockitousage/serialization/DeepStubsSerializableTest.java
Missing header in: src/test/java/org/mockitousage/junitrunner/SilentRunnerTest.java
Missing header in: src/test/java/org/mockitousage/junitrunner/UnusedStubsExceptionMessageTest.java
Missing header in: src/test/java/org/mockitousage/junitrunner/StrictRunnerTest.java
Missing header in: src/test/java/org/mockitousage/verification/DelayedExecution.java
Missing header in: src/test/java/org/mockitousage/junitrule/StubbingWarningsMultiThreadingTest.java
Missing header in: src/test/java/org/mockitousage/junitrule/StubbingWarningsJUnitRuleTest.java
Missing header in: src/test/java/org/mockitousage/junitrule/SilentJUnitRuleTest.java
Missing header in: src/test/java/org/mockitousage/junitrule/RuleTestWithParameterConstructorTest.java
Missing header in: src/test/java/org/mockitousage/junitrule/VerificationCollectorImplTest.java
Missing header in: src/test/java/org/mockitousage/junitrule/RuleTestWithFactoryMethodTest.java
Missing header in: src/test/java/org/mockitousage/junitrule/MockitoJUnitRuleTest.java
Missing header in: src/test/java/org/mockitousage/debugging/InvocationsPrinterTest.java
Missing header in: src/test/java/org/mockitousage/junitrule/InvalidTargetMockitoJUnitRuleTest.java
Missing header in: src/test/java/org/mockitousage/internal/junit/UnusedStubbingsFinderTest.java
Missing header in: src/test/java/org/mockitousage/bugs/DiamondInheritanceIsConfusingMockitoTest.java
Missing header in: src/test/java/org/mockitousage/bugs/ClassCastExOnVerifyZeroInteractionsTest.java
Missing header in: src/test/java/org/mockitousage/constructor/CreatingMocksWithConstructorTest.java
Missing header in: src/test/java/org/mockitousage/bugs/ConfusedSignatureTest.java
Missing header in: src/test/java/org/mockitousage/bugs/EqualsWithDeltaTest.java
Missing header in: src/test/java/org/mockitousage/bugs/creation/PublicMethodInParentWithNonPublicTypeInSignatureTest.java
Missing header in: src/test/java/org/mockitousage/bugs/CompareMatcherTest.java
Missing header in: src/test/java/org/mockitousage/bugs/creation/otherpackage/PublicParentClass.java
Missing header in: src/test/java/org/mockitousage/bugs/injection/Issue353InjectionMightNotHappenInCertainConfigurationTest.java
Missing header in: src/test/java/org/mockitousage/bugs/creation/ConstructorInvokingMethodShouldNotRaiseExceptionTest.java
Missing header in: src/test/java/org/mockitousage/bugs/deepstubs/DeepStubsWronglyReportsSerializationProblemsTest.java
Missing header in: src/test/java/org/mockitousage/bugs/creation/api/PublicClass.java
Missing header in: src/test/java/org/mockitousage/performance/StubOnlyAvoidMemoryConsumptionTest.java
Missing header in: src/test/java/org/mockitousage/basicapi/ResetInvocationsTest.java
Missing header in: src/test/java/org/mockitousage/bugs/deepstubs/DeepStubFailingWhenGenericNestedAsRawTypeTest.java
Missing header in: src/test/java/org/mockitousage/matchers/VarargsTest.java
Missing header in: src/test/java/org/mockitousage/stubbing/StubbingWithDelegateVarArgsTest.java
Missing header in: src/test/java/org/mockitousage/stubbing/StubbingReturnsSelfTest.java
Missing header in: src/test/java/org/mockitointegration/NoJUnitDependenciesTest.java
Missing header in: src/test/java/org/mockitoutil/ClassLoadersTest.java
Missing header in: src/test/java/org/mockitoutil/SimplePerRealmReloadingClassLoader.java
Missing header in: src/test/java/org/mockitoutil/ClassLoaders.java
Missing header in: src/test/java/org/mockitoutil/TestBaseTest.java
Missing header in: src/test/java/org/mockitoutil/VmArgAssumptions.java
Missing header in: src/test/java/org/mockitoutil/SimpleClassGenerator.java
Missing header in: src/test/java/org/mockitoutil/Stopwatch.java
Missing header in: src/test/java/org/mockitoutil/JUnitResultAssert.java
Missing header in: src/test/java/org/mockitoutil/SimpleSerializationUtil.java
Missing header in: src/test/java/org/mockito/verification/NegativeDurationTest.java
Missing header in: src/test/java/org/mockito/AnnotationsAreCopiedFromMockedTypeTest.java
Missing header in: src/test/java/org/mockito/internal/util/TimerTest.java
Missing header in: src/test/java/org/mockito/internal/util/PlatformTest.java
Missing header in: src/test/java/org/mockito/internal/util/DefaultMockingDetailsTest.java
Missing header in: src/test/java/org/mockito/internal/util/reflection/GenericArrayReturnTypeTest.java
Missing header in: src/test/java/org/mockito/internal/util/reflection/SuperTypesLastSorterTest.java
Missing header in: src/test/java/org/mockito/internal/util/reflection/GenericTypeExtractorTest.java
Missing header in: src/test/java/org/mockito/internal/verification/DescriptionTest.java
Missing header in: src/test/java/org/mockito/internal/util/io/IOUtilTest.java
Missing header in: src/test/java/org/mockito/internal/verification/VerificationOverTimeImplTest.java
Missing header in: src/test/java/org/mockito/internal/invocation/TypeSafeMatchingTest.java
Missing header in: src/test/java/org/mockito/internal/verification/VerificationWithDescriptionTest.java
Missing header in: src/test/java/org/mockito/internal/junit/JUnitRuleTest.java
Missing header in: src/test/java/org/mockito/internal/junit/StubbingArgMismatchesTest.java
Missing header in: src/test/java/org/mockito/internal/junit/UnusedStubbingsTest.java
Missing header in: src/test/java/org/mockito/internal/junit/ArgMismatchFinderTest.java
Missing header in: src/test/java/org/mockito/internal/creation/DelegatingMethodTest.java
Missing header in: src/test/java/org/mockito/internal/creation/instance/ConstructorInstantiatorTest.java
Missing header in: src/test/java/org/mockito/internal/creation/bytebuddy/AbstractByteBuddyMockMakerTest.java
Missing header in: src/test/java/org/mockito/internal/creation/bytebuddy/TypeCachingMockBytecodeGeneratorTest.java
Missing header in: src/test/java/org/mockito/internal/creation/bytebuddy/SubclassByteBuddyMockMakerTest.java
Missing header in: src/test/java/org/mockito/internal/creation/bytebuddy/InlineByteBuddyMockMakerTest.java
Missing header in: src/test/java/org/mockito/internal/configuration/ClassPathLoaderTest.java
Missing header in: src/test/java/org/mockito/internal/creation/bytebuddy/ByteBuddyMockMakerTest.java
Missing header in: src/test/java/org/mockito/internal/configuration/plugins/PluginFinderTest.java
Missing header in: src/test/java/org/mockito/internal/matchers/InstanceOfTest.java
Missing header in: src/test/java/org/mockito/internal/framework/DefaultMockitoFrameworkTest.java
Missing header in: src/test/java/org/mockito/internal/hamcrest/MatcherGenericTypeExtractorTest.java
Missing header in: src/test/java/org/mockito/internal/matchers/text/MatcherToStringTest.java
Missing header in: src/test/java/org/mockito/internal/matchers/text/ValuePrinterTest.java
Missing header in: src/test/java/org/mockito/internal/stubbing/defaultanswers/ForwardsInvocationsTest.java
Missing header in: src/test/java/org/mockito/internal/matchers/StringMatchersTest.java
Missing header in: src/test/java/org/mockito/exceptions/stacktrace/StackTraceCleanerTest.java
:license
:createTestResources
:compileTestJavaNote: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.

:removeTestResources
:processTestResources UP-TO-DATE
:testClasses
:test
:extTest:compileTestJava
:extTest:testClasses
:extTest:test
:extTest:check
:extTest:build

org.mockitousage.verification.VerificationWithTimeoutTest > shouldAllowMixingOtherModesWithTimeoutAndFail FAILED
    Wanted but not invoked:
    mock.oneArg('c');
    -> at org.mockitousage.verification.VerificationWithTimeoutTest.shouldAllowMixingOtherModesWithTimeoutAndFail(VerificationWithTimeoutTest.java:103)
    Actually, there were zero interactions with this mock.
        at org.mockito.internal.exceptions.Reporter.wantedButNotInvoked(Reporter.java:329)
        at org.mockito.internal.verification.checkers.MissingInvocationChecker.checkMissingInvocation(MissingInvocationChecker.java:38)
        at org.mockito.internal.verification.AtLeast.verify(AtLeast.java:31)
        at org.mockito.internal.verification.VerificationOverTimeImpl.verify(VerificationOverTimeImpl.java:78)
        at org.mockito.internal.verification.VerificationWrapper.verify(VerificationWrapper.java:19)
        at org.mockito.internal.verification.MockAwareVerificationMode.verify(MockAwareVerificationMode.java:21)
        at org.mockito.internal.handler.MockHandlerImpl.handle(MockHandlerImpl.java:73)
        at org.mockito.internal.handler.NullResultGuardian.handle(NullResultGuardian.java:32)
        at org.mockito.internal.handler.InvocationNotifierHandler.handle(InvocationNotifierHandler.java:36)
        at org.mockito.internal.creation.bytebuddy.MockMethodInterceptor.doIntercept(MockMethodInterceptor.java:41)
        at org.mockito.internal.creation.bytebuddy.MockMethodInterceptor$DispatcherDefaultingToRealMethod.interceptAbstract(MockMethodInterceptor.java:120)
        at org.mockitousage.IMethods$MockitoMock$442824988.oneArg(Unknown Source)
        at org.mockitousage.verification.VerificationWithTimeoutTest.shouldAllowMixingOtherModesWithTimeoutAndFail(VerificationWithTimeoutTest.java:103)

1666 tests completed, 1 failed, 48 skipped
:test FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///home/travis/build/mockito/mockito/build/reports/tests/index.html

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 1 mins 29.851 secs

Publishing build information...
https://gradle.com/s/rad43ogqfwgoi



The command "./gradlew ciBuild release -Dscan" exited with 1.
$ rm -f $HOME/.gradle/caches/modules-2/modules-2.lock

store build cache

change detected (content changed, file is created, or file is deleted):
/home/travis/.gradle/caches/2.14.1/plugin-resolution/cache.properties.lock
/home/travis/.gradle/caches/2.14.1/scripts/1q4d2xab9ze2xxke3zazxroc2/cp_dsl/cp_dsl-1646510925884990387/cache.properties
/home/travis/.gradle/caches/2.14.1/scripts/1q4d2xab9ze2xxke3zazxroc2/cp_dsl/cp_dsl-1646510925884990387/cache.properties.lock
/home/travis/.gradle/caches/2.14.1/scripts/1q4d2xab9ze2xxke3zazxroc2/cp_dsl/cp_dsl-1646510925884990387/classes/_BuildScript_.class
/home/travis/.gradle/caches/2.14.1/scripts/1q4d2xab9ze2xxke3zazxroc2/cp_dsl/cp_dsl-1646510925884990387/classes/_BuildScript_$_run_closure1.class
/home/travis/.gradle/caches/2.14.1/scripts/1q4d2xab9ze2xxke3zazxroc2/cp_dsl/cp_dsl-1646510925884990387/classes/_BuildScript_$_run_closure1$_closure2.class
/home/travis/.gradle/caches/2.14.1/scripts/1q4d2xab9ze2xxke3zazxroc2/cp_dsl/cp_dsl-1646510925884990387/classes/_BuildScript_$_run_closure1$_closure2$_closure4.class
/home/travis/.gradle/caches/2.14.1/scripts/1q4d2xab9ze2xxke3zazxroc2/cp_dsl/cp_dsl-16

...

changes detected, packing new archive
.
.
.
uploading archive


Done. Your build exited with 1.
```

--- a/src/test/java/org/mockitousage/verification/VerificationAfterDelayTest.java
+++ b/src/test/java/org/mockitousage/verification/VerificationAfterDelayTest.java
@@ -4,17 +4,6 @@
 
 package org.mockitousage.verification;
 
-import static java.util.concurrent.TimeUnit.MILLISECONDS;
-import static junit.framework.TestCase.assertEquals;
-import static org.assertj.core.api.Assertions.assertThat;
-import static org.junit.rules.ExpectedException.none;
-import static org.mockito.Mockito.after;
-import static org.mockito.Mockito.times;
-import static org.mockito.Mockito.verify;
-import static org.mockito.junit.MockitoJUnit.rule;
-import static org.mockitoutil.Stopwatch.createNotStarted;
-
-import java.util.concurrent.ScheduledExecutorService;
 import org.junit.After;
 import org.junit.Assert;
 import org.junit.Before;
@@ -27,12 +16,26 @@ import org.mockito.Mock;
 import org.mockito.exceptions.base.MockitoAssertionError;
 import org.mockito.junit.MockitoRule;
 import org.mockitousage.IMethods;
+import org.mockitoutil.RetryRule;
 import org.mockitoutil.Stopwatch;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static junit.framework.TestCase.assertEquals;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.rules.ExpectedException.none;
+import static org.mockito.Mockito.after;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.junit.MockitoJUnit.rule;
+import static org.mockitoutil.Stopwatch.createNotStarted;
 
 public class VerificationAfterDelayTest {
 
     @Rule
     public MockitoRule mockito = rule();
+
+    @Rule
+    public RetryRule retryRule = RetryRule.attempts(4);
 
     @Rule
     public ExpectedException exception = none();

--- a/src/test/java/org/mockitousage/verification/VerificationWithTimeoutTest.java
+++ b/src/test/java/org/mockitousage/verification/VerificationWithTimeoutTest.java
@@ -5,20 +5,6 @@
 
 package org.mockitousage.verification;
 
-import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
-import static java.util.concurrent.TimeUnit.MILLISECONDS;
-import static java.util.concurrent.TimeUnit.SECONDS;
-import static org.junit.rules.ExpectedException.none;
-import static org.mockito.Mockito.inOrder;
-import static org.mockito.Mockito.never;
-import static org.mockito.Mockito.timeout;
-import static org.mockito.Mockito.times;
-import static org.mockito.Mockito.verify;
-import static org.mockito.Mockito.verifyNoMoreInteractions;
-import static org.mockito.junit.MockitoJUnit.rule;
-
-import java.util.concurrent.ScheduledExecutorService;
-
 import org.junit.After;
 import org.junit.Before;
 import org.junit.Rule;
@@ -31,6 +17,17 @@ import org.mockito.exceptions.verification.NoInteractionsWanted;
 import org.mockito.exceptions.verification.TooLittleActualInvocations;
 import org.mockito.junit.MockitoRule;
 import org.mockitousage.IMethods;
+import org.mockitoutil.RetryRule;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static org.junit.rules.ExpectedException.none;
+import static org.mockito.Mockito.inOrder;
+import static org.mockito.Mockito.never;
+import static org.mockito.Mockito.timeout;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.junit.MockitoJUnit.rule;
 
 public class VerificationWithTimeoutTest {
 
@@ -38,15 +35,16 @@ public class VerificationWithTimeoutTest {
     public MockitoRule mockito = rule();
 
     @Rule
+    public RetryRule retryRule = RetryRule.attempts(4);
+
+    @Rule
     public ExpectedException exception = none();
 
     @Mock
     private IMethods mock;
 
-    private ScheduledExecutorService executor;
-
     private DelayedExecution delayedExecution;
-    
+
     @Before
     public void setUp() {
         delayedExecution = new DelayedExecution();
@@ -60,7 +58,7 @@ public class VerificationWithTimeoutTest {
     @Test
     public void shouldVerifyWithTimeout() throws Exception {
         // when
-        delayedExecution.callAsync(30, MILLISECONDS, callMock('c') );
+        delayedExecution.callAsync(30, MILLISECONDS, callMock('c'));
 
         // then
         verify(mock, timeout(100)).oneArg('c');
@@ -73,7 +71,7 @@ public class VerificationWithTimeoutTest {
     @Test
     public void shouldFailVerificationWithTimeout() throws Exception {
         // when
-        delayedExecution.callAsync(30, MILLISECONDS, callMock('c') );
+        delayedExecution.callAsync(30, MILLISECONDS, callMock('c'));
 
         // then
         verify(mock, never()).oneArg('c');
@@ -84,8 +82,8 @@ public class VerificationWithTimeoutTest {
     @Test
     public void shouldAllowMixingOtherModesWithTimeout() throws Exception {
         // when
-        delayedExecution.callAsync(10, MILLISECONDS, callMock('c') );
-        delayedExecution.callAsync(10, MILLISECONDS, callMock('c') );
+        delayedExecution.callAsync(10, MILLISECONDS, callMock('c'));
+        delayedExecution.callAsync(10, MILLISECONDS, callMock('c'));
 
         // then
         verify(mock, timeout(100).atLeast(1)).oneArg('c');
@@ -96,8 +94,8 @@ public class VerificationWithTimeoutTest {
     @Test
     public void shouldAllowMixingOtherModesWithTimeoutAndFail() throws Exception {
         // when
-        delayedExecution.callAsync(10, MILLISECONDS, callMock('c') );
-        delayedExecution.callAsync(10, MILLISECONDS, callMock('c') );
+        delayedExecution.callAsync(10, MILLISECONDS, callMock('c'));
+        delayedExecution.callAsync(10, MILLISECONDS, callMock('c'));
 
         // then
         verify(mock, timeout(100).atLeast(1)).oneArg('c');
@@ -108,7 +106,7 @@ public class VerificationWithTimeoutTest {
     @Test
     public void shouldAllowMixingOnlyWithTimeout() throws Exception {
         // when
-        delayedExecution.callAsync(30, MILLISECONDS, callMock('c') );
+        delayedExecution.callAsync(30, MILLISECONDS, callMock('c'));
 
         // then
         verify(mock, never()).oneArg('c');
@@ -118,7 +116,7 @@ public class VerificationWithTimeoutTest {
     @Test
     public void shouldAllowMixingOnlyWithTimeoutAndFail() throws Exception {
         // when
-        delayedExecution.callAsync(30, MILLISECONDS, callMock('c') );
+        delayedExecution.callAsync(30, MILLISECONDS, callMock('c'));
 
         // and when
         mock.oneArg('x');

--- a/src/test/java/org/mockitoutil/RetryRule.java
+++ b/src/test/java/org/mockitoutil/RetryRule.java
@@ -1,0 +1,53 @@
+/*
+ * Copyright (c) 2007 Mockito contributors
+ * This program is made available under the terms of the MIT License.
+ */
+
+package org.mockitoutil;
+
+import org.junit.rules.TestRule;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+public class RetryRule implements TestRule {
+    private final TestRule innerRule;
+
+    public static RetryRule attempts(final int attempts) {
+        return new RetryRule(new NumberedAttempts(attempts));
+    }
+
+    private RetryRule(TestRule innerRule) {
+        this.innerRule = innerRule;
+    }
+
+    public Statement apply(final Statement base, final Description description) {
+        return innerRule.apply(base, description);
+    }
+
+    private static class NumberedAttempts implements TestRule {
+        private final int attempts;
+
+        NumberedAttempts(int attempts) {
+            assert attempts > 1;
+            this.attempts = attempts;
+        }
+
+        @Override
+        public Statement apply(final Statement base, final Description description) {
+            return new Statement() {
+                @Override
+                public void evaluate() throws Throwable {
+                    for (int remainingAttempts = attempts; remainingAttempts > 0 ; remainingAttempts--) {
+                        try {
+                            base.evaluate();
+                        } catch (Throwable throwable) {
+                            if (remainingAttempts > 0) {
+                                throw throwable;
+                            }
+                        }
+                    }
+                }
+            };
+        }
+    }
+}

